composer/datasets/evaluator.py

```diff
@@ -0,0 +1,74 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import List
+
+import yahp as hp
+from torchmetrics.classification.accuracy import Accuracy
+from torchmetrics.classification.f_beta import F1
+from torchmetrics.classification.matthews_corrcoef import MatthewsCorrcoef
+from torchmetrics.collections import MetricCollection
+from torchmetrics.metric import Metric
+from torchmetrics.regression.pearson import PearsonCorrcoef
+from torchmetrics.regression.spearman import SpearmanCorrcoef
+
+from composer.core.types import Metrics
+from composer.datasets.dataset_registry import get_dataset_registry
+from composer.datasets.hparams import DataloaderSpec, DatasetHparams
+from composer.models.loss import CrossEntropyLoss
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class EvaluatorHparams(hp.Hparams):
+    hparams_registry = {  # type: ignore
+        "eval_dataset": get_dataset_registry(),
+    }
+
+    metric_registry = {
```

Comment:
Ideally, we want to support all metrics listed in https://torchmetrics.readthedocs.io/en/latest/references/modules.html (there are a lot) in addition to custom metrics that the user has defined (see `Dice` in `composer/models/loss.py` as an example).

We shouldn't have to instantiate the metrics in Hparams because each model class that extends `BaseMosaicModel` returns instantiated objects for each metric that can be computed for the model. 

The evaluator only needs to track the names of the metrics so that it can filter out metrics that shouldn't be used for the given eval_dataset. 

For example, we can have `metric_names: List[str]` be a field here and then in the `eval()` loop of the `Trainer` we can check the `metric_names` against the metrics returned by `model.metrics()` and only compute the metrics whose names match the ones for this `Evaluator`.

Reply:
Oh also another thing - the list of metric names should be optional so that if not provided it defaults to using all metrics returned by `model.metrics()`